sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -22,5 +24,183 @@ public partial class LogsIngestionClient
+            uri.AppendPath(ruleId, true);
+            uri.AppendPath("/streams/", false);
+            uri.AppendPath(streamName, true);
+            uri.AppendQuery("api-version", _apiVersion, true);
+            request.Uri = uri;
+            request.Headers.Add("Accept", "application/json");
+            request.Headers.Add("Content-Type", "application/json");
+            if (contentEncoding != null)
+            {
+                request.Headers.Add("Content-Encoding", contentEncoding);
+            }
+            if (contentEncoding == "gzip")
+            {
+                GZipUtf8JsonRequestContent gzContent = new(content);
+                request.Content = gzContent;
+            }
+            else
+            {
+                request.Content = content;
+            }
+            return message;
+        }
+
+        /// <summary>
+        /// Hidden method for batching data
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="logEntries"></param>
+        /// <returns></returns>
+        internal static IEnumerable<BinaryData> Batching<T>(IEnumerable<T> logEntries)
+        {
+            //TODO: use Array pool instead
+            MemoryStream stream = new MemoryStream(_singleUploadThreshold);
+            WriteMemory(stream, BinaryData.FromString("[").ToMemory());
+            int entryCount = 0;
+            foreach (var log in logEntries)
+            {
+                BinaryData entry = log is BinaryData d ? d : BinaryData.FromObjectAsJson(log);
+
+                var memory = entry.ToMemory();
+                if ((stream.Length + memory.Length + 1) >= _singleUploadThreshold) // if adding this entry makes stream > 1 Mb send current stream now
+                {
+                    WriteMemory(stream, BinaryData.FromString("]").ToMemory());
+                    stream.Position = 0; // set Position to 0 to return everything from beginning of stream
+                    yield return (BinaryData.FromStream(stream));
+                    stream = new MemoryStream(_singleUploadThreshold); // reset stream
+                }
+                else
+                {
+                    WriteMemory(stream, memory);
+                    WriteMemory(stream, BinaryData.FromString(",").ToMemory());
+                    if ((entryCount + 1) == logEntries.Count())
+                    {
+                        // reached end of logEntries and we haven't returned yet
+                        WriteMemory(stream, BinaryData.FromString("]").ToMemory());
+                        stream.Position = 0;
+                        yield return (BinaryData.FromStream(stream));
+                    }
+                }
+                entryCount++;
+            }
+        }
+
+        private static void WriteMemory(MemoryStream stream, ReadOnlyMemory<byte> memory)
+        {
+            stream.Write(memory.ToArray(), 0, memory.Length); //TODO: fix ToArray
```

Comment:
I think `memory.ToArray()` is allocating here.  Can you try `memory.AsSpan()` instead?